cozy/structures/ordered.py

```diff
@@ -0,0 +1,203 @@
+'''Ordered multiset structure
+TODO(zhen): define a virtual class for universal structure extension
+            and move useful documentation there from heaps.py and this file
+'''
+
+from cozy.common import pick_to_sum, No
+from cozy.syntax import *
+from cozy.target_syntax import SInsert, SErase, TOrderedNative
+from cozy.syntax_tools import fresh_var, pprint
+from cozy.pools import Pool, RUNTIME_POOL, STATE_POOL
+
+TMinOrdered = declare_case(Type, "TMinOrdered", ["elem_type"])
+TMaxOrdered = declare_case(Type, "TMaxOrdered", ["elem_type"])
+EMakeMaxOrdered = declare_case(Exp, "EMakeMaxOrdered", ["e"])
+EMakeMinOrdered = declare_case(Exp, "EMakeMinOrdered", ["e"])
+def with_type_safe_min(self, t):
+    assert isinstance(t, TMinOrdered)
+    self.type = t
+    return self
+def with_type_safe_max(self, t):
+    assert isinstance(t, TMaxOrdered)
+    self.type = t
+    return self
+EMakeMaxOrdered.with_type = with_type_safe_max
+EMakeMinOrdered.with_type = with_type_safe_min
+
+EOrderedElems = declare_case(Exp, "EOrderedElems", ["e"])
+EOrderedPeek  = declare_case(Exp, "EOrderedPeek", ["e", "index"])
+
+def _parent(idx : Exp) -> Exp:
+    return EBinOp(idx, "-", ONE).with_type(INT)
+
+class Ordereds(object):
```

Comment:
I am not sure that `Ordered` is a good name. I would call this `TreeMultisets` (along with nodes like `TMinTreeMultiset` and `EMakeMaxTreeMultiset` and so forth). (Similarly, I now wish I had named `Heaps` as `ArrayHeaps`.)